File: aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.updater;

import android.text.TextUtils;

import com.aliucord.*;
import com.aliucord.api.NotificationsAPI;
import com.aliucord.entities.NotificationData;
import com.aliucord.entities.Plugin;
import com.aliucord.settings.AliucordPageKt;
import com.aliucord.utils.MDUtils;
import com.google.gson.reflect.TypeToken;

import java.io.File;
import java.lang.reflect.Type;
import java.util.*;
import java.util.concurrent.TimeUnit;

import kotlin.Unit;

public class PluginUpdater {
    public static class UpdateInfo {
        public int minimumDiscordVersion;
        public String version;
        public String build;
        public String changelog;
        public String changelogMedia;
        public String sha1sum;
    }

    public static class CachedData {
        public Map<String, UpdateInfo> data;
        public long time = System.currentTimeMillis();

        public CachedData(Map<String, UpdateInfo> d) { data = d; }
    }

    public static final Logger logger = new Logger("Updater");

    // Synchronized to avoid ConcurrentModificationException
    public static final Map<String, CachedData> cache = Collections.synchronizedMap(new HashMap<>());
    public static final Map<String, String> updated = Collections.synchronizedMap(new HashMap<>());
    public static final List<String> updates = Collections.synchronizedList(new ArrayList<>());

    private static final Type resType = TypeToken.getParameterized(Map.class, String.class, UpdateInfo.class).getType();

    public static void checkUpdates(boolean notify) {
        updates.clear();
        for (Map.Entry<String, Plugin> plugin : PluginManager.plugins.entrySet()) {
            if (checkPluginUpdate(plugin.getValue()))
                updates.add(plugin.getKey());
        }
        if (!notify || (updates.size() == 0 && !(!Updater.usingDexFromStorage() && Updater.isAliucordOutdated()))) return;

        NotificationData notificationData = new NotificationData()
                .setTitle("Updater")
                .setAutoDismissPeriodSecs(10)
                .setOnClick(view -> {
                    Utils.openPage(Utils.appActivity, com.aliucord.settings.Updater.class);
                    return Unit.a;
                });

        String updatablePlugins = String.format("**%s**", TextUtils.join("**, **", updates.toArray()));
        String body;
        if (Main.settings.getBool(AliucordPageKt.AUTO_UPDATE_PLUGINS_KEY, false)) {
            int res = PluginUpdater.updateAll();
            if (res == 0) return;
            if (res == -1) {
                body = "Something went wrong while auto updating plugins. Check the debug log for more info.";
            } else {
                body = String.format("Automatically updated %s: %s", Utils.pluralise(res, "plugin"), updatablePlugins);
            }
        } else if (updates.size() != 0) {
            body = "Updates for plugins are available: " + updatablePlugins;
        } else body = "All plugins up to date!";

        if (!Updater.usingDexFromStorage()) {
            if (Updater.isDiscordOutdated()) {
                body = "Your Base Discord is outdated. Please update using the installer - " + body;
            } else if (Updater.isAliucordOutdated()) {
                if (Main.settings.getBool(AliucordPageKt.AUTO_UPDATE_ALIUCORD_KEY, false)) {
                    try {
                        Updater.updateAliucord(Utils.appActivity);
                        body = "Auto updated Aliucord. Please restart Aliucord to load the update - " + body;
                    } catch (Throwable th) {
                        body = "Failed to auto update Aliucord. Please update it manually - " + body;
                    }
                } else {
                    body = "Your Aliucord is outdated. Please update it to the latest version - " + body;
                }
            }
        }

        notificationData.setBody(MDUtils.render(body));
        NotificationsAPI.display(notificationData);
    }

    public static boolean checkPluginUpdate(Plugin plugin) {
        Plugin.Manifest manifest = plugin.getManifest();
        if (manifest.updateUrl == null || manifest.updateUrl.equals("")) return false;

        try {
            UpdateInfo updateInfo = getUpdateInfo(plugin);
            if (updateInfo == null || updateInfo.minimumDiscordVersion > Constants.DISCORD_VERSION) return false;

            String updatedVer = updated.get(plugin.getClass().getSimpleName());
            if (updatedVer != null && !Updater.isOutdated(plugin.getName(), updateInfo.version, updatedVer)) return false;

            return Updater.isOutdated(plugin.getName(), manifest.version, updateInfo.version);
        } catch (Throwable e) { logger.error("Failed to check update for: " + plugin.getClass().getSimpleName(), e); }
        return false;
    }

    public static UpdateInfo getUpdateInfo(Plugin plugin) throws Exception {
        Plugin.Manifest manifest = plugin.getManifest();
        if (manifest.updateUrl == null || manifest.updateUrl.equals("")) return null;
        String name = plugin.getName();

        CachedData cached = cache.get(manifest.updateUrl);
        if (cached != null && cached.time > System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(30)) {
            UpdateInfo updateInfo = cached.data.get(name);
            UpdateInfo defaultInfo = cached.data.get("default");
            if (updateInfo == null) return defaultInfo;
            addDefaultInfo(updateInfo, defaultInfo);
            return updateInfo;
        }

        Map<String, UpdateInfo> res = Http.simpleJsonGet(manifest.updateUrl, resType);
        if (res == null) return null;
        cache.put(manifest.updateUrl, new CachedData(res));
        UpdateInfo updateInfo = res.get(name);
        UpdateInfo defaultInfo = res.get("default");
        if (updateInfo == null) return defaultInfo;
        addDefaultInfo(updateInfo, defaultInfo);
        return updateInfo;
    }

    public static int updateAll() {
        int updateCount = 0;
        for (String plugin : updates) {
            try {
                if (update(plugin) && updateCount != -1) updateCount++;
            } catch (Throwable t) {
                logger.error("Error while updating plugin " + plugin, t);
                updateCount = -1;
            }
        }
        updates.clear();
        checkUpdates(false);
        return updateCount;
    }

    public static boolean update(String plugin) throws Throwable {
        var p = PluginManager.plugins.get(plugin);
        if (p == null)
            throw new NoSuchElementException("No such plugin: " + plugin);

        var updateInfo = getUpdateInfo(p);
        if (updateInfo == null) return false;

        var url = updateInfo.build.replace("%s", plugin);

        try (var res = new Http.Request(url).execute()) {
            res.saveToFile(
                    new File(Constants.PLUGINS_PATH + "/", p.__filename + ".zip"),
                    updateInfo.sha1sum
            );
        }

        if (PluginManager.isPluginEnabled(plugin)) {
            Utils.mainThread.post(() -> {
                PluginManager.remountPlugin(plugin);
                var newPlugin = Objects.requireNonNull(PluginManager.plugins.get(plugin));
                if (newPlugin.requiresRestart()) Utils.promptRestart();
            });
        }
        updated.put(plugin, updateInfo.version);
        return true;
    }

    private static void addDefaultInfo(UpdateInfo updateInfo, UpdateInfo defaultInfo) {
        if (defaultInfo != null) {
            if (updateInfo.minimumDiscordVersion == 0) updateInfo.minimumDiscordVersion = defaultInfo.minimumDiscordVersion;
            if (updateInfo.version == null) updateInfo.version = defaultInfo.version;
            if (updateInfo.build == null) updateInfo.build = defaultInfo.build;
        }
    }
}

File: Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.kt
package com.aliucord.updater

import android.os.Build
import com.aliucord.*
import com.aliucord.Utils.openPage
import com.aliucord.api.NotificationsAPI
import com.aliucord.entities.*
import com.aliucord.screens.UpdaterScreen
import com.aliucord.settings.AUTO_UPDATE_PLUGINS_KEY
import com.aliucord.utils.MDUtils
import com.aliucord.utils.SemVer
import java.io.File

/**
 * Manages fetching and installing plugin updates.
 * This class is pure and does not store plugin updates globally.
 */
internal object PluginUpdater {
    private val logger = Logger("Updater/Plugins")

    /**
     * Represents an available plugin update.
     */
    data class PluginUpdate(
        /**
         * The currently loaded plugin this is update applies to.
         */
        val plugin: Plugin,
        /**
         * The plugin's manifest name/id
         */
        val pluginName: String = plugin.name,
        /**
         * The fetched update info for the latest build of this plugin.
         */
        val info: PluginUpdaterSource.PluginBuildInfo,
    ) {
        /**
         * Whether the base Discord/Aliucord installation is outdated and
         * requires a reinstallation update through Aliucord Manager.
         */
        val isBaseOutdated: Boolean = info.minimumDiscordVersion > Constants.DISCORD_VERSION ||
            !ManagerBuild.hasKotlin(info.minimumKotlinVersion.toString())

        /**
         * Whether the current Aliucord core is outdated and requires an update.
         */
        val isCoreOutdated: Boolean = (info.minimumAliucordVersion ?: SemVer.Zero) > SemVer.parse(BuildConfig.VERSION)

        /**
         * Whether the current Android version is too low to load the new plugin.
         */
        val isAndroidOutdated: Boolean = info.minimumApiLevel > Build.VERSION.SDK_INT

        /**
         * Whether this plugin should be allowed to update as
         * the new build will not cause issues upon loading.
         */
        fun isUpdatePossible(): Boolean =
            !isBaseOutdated && !isCoreOutdated && !isAndroidOutdated
    }

    /**
     * Determines whether automatic plugin updates have been enabled by the user.
     */
    @JvmStatic
    fun isAutoUpdateEnabled(): Boolean = Main.settings.getBool(AUTO_UPDATE_PLUGINS_KEY, false)

    /**
     * Force fetches all available updates, including ones that cannot be updated.
     * The resulting updates should not be held for long durations (ie, cached globally).
     */
    @JvmStatic
    fun fetchUpdates(source: PluginUpdaterSource): List<PluginUpdate> {
        logger.info("Checking for plugin updates...")

        val updates = mutableListOf<PluginUpdate>()
        for (plugin in PluginManager.plugins.values) {
            try {
                if (plugin is CorePlugin) continue

                val info = source.getPluginBuildInfo(
                    pluginName = plugin.manifest.name ?: continue,
                    updateInfoUrl = plugin.manifest.updateUrl
                        ?.takeIf { it.isNotEmpty() }
                        ?: continue,
                )
                // Previous attempt at retrieving updater data failed
                if (info == null) {
                    logger.warn("Failed to check updates for plugin ${plugin.name} (${plugin.__filename}.zip)")
                    continue
                }

                // Assume invalid local versions are always out of date
                val localVersion = SemVer.parseOrNull(plugin.manifest.version) ?: SemVer.Zero

                // Plugin is already up-to-date
                if (localVersion >= info.version)
                    continue

                updates += PluginUpdate(
                    plugin = plugin,
                    pluginName = plugin.name,
                    info = info,
                )
            } catch (e: Exception) {
                logger.error("Failed checking updates for plugin ${plugin.name} (${plugin.__filename}.zip)", e)
                continue
            }
        }
        return updates
    }

    @JvmStatic
    fun startupCheck() {
        val updates = fetchUpdates(PluginUpdaterSource())
        if (updates.isEmpty()) return

        // Only show update notification
        if (!isAutoUpdateEnabled()) {
            val notification = NotificationData()
                .setTitle("Updater")
                .setBody(MDUtils.render(buildString {
                    append("Found ${updates.size} available plugin updates: ")
                    append(updates
                        .take(5)
                        .joinToString(separator = ", ", transform = { "**${it.pluginName}**" }))

                    if (updates.size > 5)
                        append(", and ${updates.size - 5} others.")

                    append(". Click to see more.")
                }))
                .setAutoDismissPeriodSecs(30)
                .setOnClick { openPage(Utils.appActivity, UpdaterScreen::class.java) }

            NotificationsAPI.display(notification)
            return
        }

        // Update plugins
        val (succeeded, failed) = updates
            .filter(PluginUpdate::isUpdatePossible)
            .partition { updatePlugin(it) }

        val notification = if (failed.isNotEmpty()) {
            NotificationData()
                .setTitle("Updater")
                .setAutoDismissPeriodSecs(30)
                .setBody(MDUtils.render(buildString {
                    append("Failed to update some plugins: ")
                    append(failed
                        .take(5)
                        .joinToString(separator = ", ", transform = { "**${it.pluginName}**" }))
                    if (failed.size > 5)
                        append(", and ${failed.size - 5} others.")
                }))
                .setOnClick { openPage(Utils.appActivity, UpdaterScreen::class.java) }
        } else {
            NotificationData()
                .setTitle("Updater")
                .setAutoDismissPeriodSecs(10)
                .setOnClick { /* Do nothing */ }
                .setBody(MDUtils.render(buildString {
                    append("Automatically updated plugins: ")
                    append(succeeded
                        .take(5)
                        .joinToString(separator = ", ", transform = { "**${it.pluginName}**" }))
                    if (succeeded.size > 5)
                        append(", and ${succeeded.size - 5} others.")
                }))
        }
        NotificationsAPI.display(notification)
    }

    @JvmStatic
    fun updatePlugin(update: PluginUpdate): Boolean {
        if (!update.isUpdatePossible())
            throw IllegalArgumentException("Cannot perform plugin update that is not possible")

        return try {
            // Legacy build url style, which contains %s as a placeholder for the plugin name (optional now)
            val downloadUrl = update.info.buildUrl.replace("%s", update.pluginName)

            Http.Request(downloadUrl).execute().use { resp ->
                // TODO: verify crc32
                resp.saveToFile(File(Constants.PLUGINS_PATH, "${update.plugin.__filename}.zip"))
            }

            reloadPlugin(update.plugin)
            true
        } catch (e: Exception) {
            logger.error("Failed to update plugin ${update.plugin} (${update.plugin.__filename}.zip)", e)
            false
        }
    }

    private fun reloadPlugin(plugin: Plugin) {
        // FIXME: plugin not reloaded when disabled
        if (!PluginManager.isPluginEnabled(plugin.name)) return

        Utils.mainThread.post {
            PluginManager.remountPlugin(plugin.name)
            val newPlugin = PluginManager.plugins[plugin.name] ?: return@post

            // FIXME: only prompt once when updating all plugins at once
            if (plugin.requiresRestart() || newPlugin.requiresRestart()) {
                Utils.promptRestart("Plugin update requires a restart. Restart now?")
            }
        }
    }
}
